### DISCUSSION TITLE
Easier import of page content

### DISCUSSION THREAD

**Discussion by ProQuestionAsker** [pngwn/MDsveX#299] [Ideas]
I'm working on creating a blog using SvelteKit and MDsveX and I'd like to be able to take small sections from one post and essentially inject them into the content of another post. I'm trying to create a Svelte-y version of [this](https://ncase.me/nutshell-wip/try/).

There are a few existing features that are similar:
* The ability to [import frontmatter](https://mdsvex.com/docs#frontmatter-1)
* The ability to [import markdown files as components](https://mdsvex.com/docs#extensions)

But what would be really helpful is to also be able to import either the body of a document or select sections of that document's body. 

Something like:
```
import { body } from "./some-mdsvex-file.svx";
```

Would probably be a great first step!

Something like this would solve my problem, but I think could also be used to solve a few other noted issues folks have been running into like creating a [list of header elements](https://github.com/pngwn/MDsveX/discussions/268) for a table of contents or [fetching all content](https://github.com/pngwn/MDsveX/issues/294) from files.

**Comment by pngwn** [comment]
What would this named import (`content`) actually contain in this case?

**Reply by ProQuestionAsker** [reply]
Ideally, any content outside the front matter. If I'm thinking of this correctly, being able to import all of the non-front matter content would allow users to parse the resulting `content` in whatever way they need (e.g., extract just the headers, split into sections of content etc.)

So like if the markdown file is like this:

```md
---
title: my awesome post
author: me
---
 
## More content
The rest of my content is all down here. This goodness and the associated headers are what I want.
```

I should note that I also am pretty new to how MDsveX is working "under the hood", so I don't actually know if this is possible.

**Reply by pngwn** [reply]
Ah, I see. So it would be the original source. Interesting. (Correct me if I'm wrong here, I'm assuming you mean the original source rather than the compiled html/ svelte). 

This is technically possible (and relatively straightforward actually) and I think the performance cost would be zero because it would be treeshakeable, so it would only get bundled if it were imported. 

My only concern would be import name collisions, there are already a few special imports having too many could cause some confusion for users as mdsvex documents are really just svelte files in disguise, so you can still define your own exports. I doubt one more standard import would hurt too much though. 

In the case of pulling out headings etc, I really do think that plugins are the way to go rather than something like this. There is a learning curve but it is much simpler as you already have access to the raw text and the heading level for example.

**Reply by ProQuestionAsker** [reply]
I think the original source would make the most sense? But, I’m still wrapping my brain around when things are processed. 

So, ultimately, I’d like to have a button in one post that when clicked, shows a specific section from a different post. 

I think the simplest way to do that would be to create the button component that can import the source material from a different post and trim it down to just the section I need. But if I’m passing this through a component and I’m using the original markdown source, will it still be rendered by MDsveX? Or will it result in “raw” markdown?

To make this all more difficult, I’m planning on also making this a static site… so, what I’m trying to do really may not be easily doable. 

Definitely hear you on the header one though. The solution was to write a plug-in that would inject the headers into the front matter, right?

**Reply by pngwn** [reply]
It would be 'raw source', so it would be equivalent to:

```js
export const content = `
## title

some paragraph
`;
```

Alongside the other stuff.

If you meant that it should be the processed source, then it is more challenging. Svelte doesn't have a way to export 'parts' of a file from a component. Take something like this:

```markdown
<script>
  let name = 'Amber';
</script>

# title

My name is {name}.
```

If we wanted _only_ the markup (and not the script), I'm not sure how it should be handled because without the script the content doesn't make any sense. But even when we say _only_ markup we have to be clear about what this means. 

Does it mean that it should be treated as if we cut out a piece of the file and then run it through the mdsvex compiler? In this case the above would break because `name` would not be defined anywhere. Does it mean we should pull out parts in a way that mean they can still function (so also pulling out the `name` definition to keep things functioning)? This would be very complex, we can't just chop things up because `name` could be used several times in a file. Do we mean that is should be that part of the file _after_ it has been compiled and turned into _static_ HTML (so this would be `<h1>title</h1> <p>My name is Amber</p>`)? I think this would be almost impossible for mdsvex because it runs very early on in the process, the static HTML generation (which is entirely optional anyway) runs at the very end.

Processing an mdsvex file from raw source to static HTML looks something like this:

**This is the mdsvex part, mdsex is a preprocessor so does its work ahead of svelte**

- mdsvex document
- Abstract Syntax Tree of mdsvex (essentially just a bit object representing the structure)
- Svelte component (the HTML version, the kind you would normally write yourself)

Then the svelte compiler takes over:

- Svelte component original source 
- Svelte component JS module. This would be a class for the component itself, any 'exports' are named exports of any value.

At this point there is a 'component' that can be executed (either in the browser or in node etc, DOM compilation vs SSR compilation). To transform this into static HTML we need the SSR version and then we would 'run' that function with the necessary props which would produce a string of HTML that we can then write to file.

This is basically what we are doing when we go from mdsvex to static HTML (via svelte kit etc). We knit this together by passing `mdsvex` as a preprocessor to the svelte compiler. It will run all `svx` documents through `mdsvex` producing valid svelte code before compiling them. In static or prerender mode, svelte kit will also take care of the last two parts for you, compiling it to and SSR component and then executing that component and writing the result to an HTML file.

The output at different stages look like this:

Raw mdsvex: 


```markdown
# title

I am a cucumber
```

mdsvex AST, this is just a massive object representing the structure of the document. This is what you have to work with if you write a 'plugin' as I was discussing. This is why plugins are good for many cases, because they have access to a lot of information already without needing to do any additional work.

```js
const AST = {
  type: 'root',
  children: [
    {
      type: 'heading', 
      level: 1, 
      children: [{
        type: 'text',
        value: 'title'
      }]
    }, {
      type: 'paragraph', 
      children: [{
        type: 'text',
        value: 'I am a cucumber'
      }]
    }
  ]
}
```

Svelte component source:

```svelte
<h1>title</h1>
<p>I am a cucumber</p>
```

A svelte js module can take one of two forms. There are DOM and SSR versions. This is pseudocode but DOM might be like this:

```js
export default class extends SvelteComponent {
  create(props) {
    const h1 = document.createElement('h1');
    h1.textContent = 'title';
    const p = document.createElement('p');
    p.textContent = 'I am a cucumber';
  }
}
```

An SSR version of this same component would look more like this:

```js
export default function(props) {
  return `<h1>title</h1><p>I am a cucumber<p>`;
}
```

You can see the actual code Svelte generates [in this REPL](https://svelte.dev/repl/3466fa1ddf2c4a70bb1655cff19ad6de?version=3.42.4) (just click the `JS` tab on the right).

These things can then be run to generate either static html (for the SSR component) which svelte kit will write to a file (which you can then load in a browser rendering the page tot he DOM), or the DOM version can be run in the browser to render actual DOM to the page.

**Reply by pngwn** [reply]
Now I got lost there but wanted to explain the pipeline a little to demonstrate why some kinds of inclusion are really simple but others aren't.

One question I do have, regarding injecting content, is why not make that a reusable chunk, like you would with a component. So whichever 'piece' of a document you wish to share, create a new mdsvex file for it and then you use that in _both_ places it is needed like this: `<SomeContent />`.

I do feel like i may still be missing something but I'd be curious to see a more concrete example if you had one, so I can get my head around it (being able to see the contents of each of the files at play, which parts would need to extracted and injected into another file, and what the final output/ outcome should look like).

**Reply by ProQuestionAsker** [reply]
First, thank you SO MUCH for typing out that extensive explanation. That helps me immensely to figure out what I'm trying to accomplish and where the right parts to jump in may be. You bring up excellent arguments about possibly having components inside of the content of a markdown file that may be without reference if simply plucked out. I hadn't thought about that but you're definitely right that that would be problematic. 

I could see how your workaround to have a specific file like `<SomeContent />` could work, but I'm not sure if it would actually work for my purposes. I'll do my best to describe my use-case.

As I mentioned, I'm working on creating a blog. If I'm writing a blog post about how to use `mdsvex`, I may mention a related concept like `markdown`. It may be outside the scope of _this_  mdsvex post to explain a bunch of intricacies about markdown itself. But, if I have already written another article about markdown, I could theoretically re-use some of that content to provide extra context for readers. So, I would have some prose, and where I would normally just link to my markdown blogpost from inside the prose of the mdsvex post, I'd use a custom component instead of an `<a>` link. When clicked, this custom component would display some of the content from the post about markdown. I'm hoping this would help users to dig deeper and find more context without having to navigate away from the article they were reading. 

Nicky Case has a [functional beta implementation](https://ncase.me/nutshell-wip/try/) of this in vanilla js that I think illustrates the purpose well. The example that says "From Other Page" is ultimately what I'm getting stuck on trying to recreate. I could just build off of Nicky's example, but their example doesn't have the benefit of having all of the content that I'll want to pull from all within the same project. I (naively) thought that working within my SvelteKit setup would actually simplify this process. 

So your suggestion to use a component like `<SomeContent />` could work, but it would probably break the flow of my writing since it would exist outside of the "original" blog post.  And could get tricky at scale if I use this for lots of references throughout my blog. 

It's quite possible that what I'm trying to make is _way_ outside the scope of mdsvex. But since it's markdown content I'm trying to use in multiple places, I figured mdsvex may have a way to simplify this.

**Reply by pngwn** [reply]
Ah, I understand now. When i originally looked at the link you posted, I didn't look properly. This makes sense now.

I also appreciate that the separate component option wouldn't work here as you don't know ahead of time what you might want to embed in future articles.

This is very reminiscent of [Andy Matuschak's 'digital garden'](https://notes.andymatuschak.org/About_these_notes), although he is linking pages rather than segments, and those pages are notes rather than articles.

> It's quite possible that what I'm trying to make is way outside the scope of mdsvex. But since it's markdown content I'm trying to use in multiple places, I figured mdsvex may have a way to simplify this.

Maybe but it is sometimes i am _very_ interested in, it was one of the use cases I was designing for in when I created mdsvex, so it is interesting that it has come up, albeit in a slightly different form.

When I wrote earlier about the pipeline for mdsvex, I skimmed over a few things. the import part is before it gets to svelte, so I'll focus there. mdsvex gets transfromed in a few passes:

- -> Source code (mdsvex)
- -> mdsvex AST (the object i listed above)
- -> HTML AST (different object representing HTML/Svelte)
- -> Svelte source code

If we were to do this using some syntax like `file#some section` and it only existed in the markdown then we'd probably have to go and read some files or something, which isn't really something we want to be doing. We don't want to emulate a bundler here as it is complex and error-prone.

So we want to use imports as much as possible because then the bundler can handle the rest. If we tried to export these 'sections' as source code (as I initially auggested) then we have a slightly different problem but still not one we want to solve. In this case we'd have fragments of source that we need to compile to Svelte or HTML for them to be usable.

What we really want is the final Svelte/HTML that is generated to be exported _somehow_. The important part of the pipeline i mentioned above are the AST parts.

This part is not really useful to us, as it represents markdown:

```js
const AST = {
  type: 'root',
  children: [
    {
      type: 'heading', 
      level: 1, 
      children: [{
        type: 'text',
        value: 'title'
      }]
    }, {
      type: 'paragraph', 
      children: [{
        type: 'text',
        value: 'I am a cucumber'
      }]
    }
  ]
}
```

However, this gets converted into _another_ AST that represents HTML (or Svelte):

```js
const AST = {
  type: 'root',
  children: [
    {
      type: 'element', 
      tagName: 'h1', 
      children: [{
        type: 'text',
        value: 'title'
      }]
    }, {
      type: 'element',
      tagName: 'paragraph', 
      children: [{
        type: 'text',
        value: 'I am a cucumber'
      }]
    }
  ]
}
```

This _is_ useful because we can actually compile this to HTML with some utilities. 

What I'm not certain of right now, is how do we expose it, because Svelte doesn't work by importing/ exporting strings. What we could do is create a string version of the sections. So this:

```markdown
---
title: my awesome post
author: me
---
 
## First section

Amazing paragraph.

## Second section

Another amazing paragraph.
```

Could be turned into this:

```html
<script content=module>
  export const metadata = { title: 'my awesome post', author: 'me' };
  
  // not sure about naming
  export const section_first_section = `<p>Amazing paragraph.</p>`;
  export const section_second_section = `<p>Another amazing paragraph.</p>`;
</script>

<h2>First Section</h2>
<p>Amazing paragraph.</p>

<h2>Second Section</h2>
<p>Another amazing paragraph.</p>
```

Now this component is essentially sitting there, waiting for you to import whatever 'section' of content you need. Now this is currently a string of html, so we couldn't use it as a component but we could use it with svelte's `{@html}` tag or pass it _to_ a component. One limitation of this is that it would only work with 'static' markdown. Well, it would work for anything but everything is being turned into a string, so any curly braces, for example would be curly braces; they wouldn't be dynamic anymore. 

In your markdown file you could do something like:
```markdown
<script>
  import { Nutshell } from './somewhere';
  import { section_first_section } from './my-file.svx';
</script>

## Just a normal article

A paragraph.

<Nutshell content={section_first_section} />
```

At its simplest `Nutshell` could be:

```svelte
<script>
  export let content;
</script>
  
{@html content}
```

There may be a way to make this work any content but that would be much more involved.

**Reply by pngwn** [reply]
You could also define some custom markdown syntax:

`:nutshell[section_first_section]` could produce `{@html section_first_section}`

**Reply by ProQuestionAsker** [reply]
Yes! So, what I'm trying to build actually _is_ a digital garden (I just refer to it as a "blog" sometimes since that term seems to be more accessible). It's very much like a minimal, inline version of what Andy Matuschak has created. Sort of like the "tooltips" you get in [Maggie Appleton's digital garden](https://maggieappleton.com/garden) when you hover over any of the orange (i.e., internal) links. 

Anyway, what you're describing is _exactly_ what I'm interested in. If I'm understanding this correctly, the limitations would be that any Svelte component execution that happens within the imported section will be turned into a string rather than executed. So, something like this: 

```markdown
---
title: my awesome post
author: me
---

<script>
import Boinger from '$lib/Boinger.svelte'
</script>
 
## First section

Amazing paragraph.

## Second section

Another amazing paragraph.
<Boinger />

```

mdsvex would normally parse into a Svelte component where `Boinger` actually does what is expected (in the case of the mdsvex playground, `Boinger` rearranges numbered blocks). But if we were to import that second section like this: `<Nutshell content={section_second_section} />` it would render `<Boinger \>` as a string rather than interactive, rearranging numbered blocks. Am I understanding that correctly?

**Reply by pngwn** [reply]
Yeah that's right. Which is obviously a bit of a pain, if you created some interactive widgets or charts, for example, it would be really nice to be able to include those as well as any simple HTML. It would definitely be easier to do if mdsvex has more constraints, I think checking for components and injecting their imports is pretty reasonable (although slightly involved) but dealing with all of Svelte's other syntax is probably not realistic. 

Originally mdsvex was just markdown + components, I have plans to create a slimmed down version of mdsvex that is exactly this, when v1 is done, partly for this reason. With more constraints there is actually a lot more you can do. Features like this one be some more manageable and there are more optimisations mdsvex (and other related tooling) could perform. 

For now, a static version of this seems possible at least. Although I'm sure there are a bunch of edge cases I haven't considered. 

---

I'm also very interested in (borderline obsessed with) digital gardens, Zettelkasten, bidirectional hypertext, etc. So do share anything you come up with, I'd be very excited to see the results!

If you decide to have a go at implementing something like the above, feel free to give me a shout if you need any help with anything. I think it can be implemented as a remark or rehype plugin today (happy to give pointers on how they work, the docs for the unified ecosystem are lacking) but v1 will change a bunch of stuff regarding plugins. If not, I'll have a look myself when I find some time (probably after v1 which is in progress).

**Reply by ProQuestionAsker** [reply]
Awesome! I don't see the "stringifying" of components being a huge issue for me at the moment, but definitely a good thing to be aware of. I may start poking around at remark or rehype plugins and see if I can get something working. I'll definitely keep you posted!

--- 
As far as digital gardens/personal knowledge management systems go, I'm with you! I stumbled upon them like 1.5 months ago and ended up going down some pretty deep rabbit holes learning about it. I feel like I've been looking for systems like this for so long! 

When I have my page up and running, I'd be happy to send it your way. I'm building it on a kind of... perhaps weird tech stack? lol I'm using [Notion](https://notion.so) as a CMS of sorts. In part because I was already organizing my personal notes and such in Notion, and because Notion also now allows backlinks between pages and has an API. So I'm converting Notion -> markdown -> Svelte (via SvelteKit and mdsvex) in what is probably the most roundabout way to handle all of this ever 😆  

I could realistically skip the markdown process step and go straight from Notion API -> SvelteKit, but since both are still in beta, I like knowing that if I have markdown files in the middle ground, I can easily switch away from Notion without breaking the whole frontend experience. 

Anyways, have you created any sort of digital garden/similar system of your own?

**Reply by pngwn** [reply]
Ah interesting, I've heard a few people going the notion direction. I've always been a little put off by it due to the performance of the app itself but it makes sense, having a ready-made editor. I'd love to see what you come up with.

I haven't done anything myself yet, I've been doing a lot of research and I've tried a lot of the existing tools that are out there. I have a few issues with all of them. I think something I'm working through at the minute is combining all of the 'parts' that make up the thing. I have seen good solutions for some of those parts but never the whole.

For me there are a few pieces to it. 

- There is the bit that makes it a garden or a Zettelkasten or a wiki. The mechanics of how pages and notes are related to one another. This more theoretical but an important part of the puzzle as it related to how you would actually use such a tool. 
- There is process of actually defining those relationships (different tools take vastly different approaches here). 
- There is the actual application itself, how do we keep track of these relationships on a page by page basis in a way that is lightweight and powerful enough to be useful? We don't want to load the metadata for thousands of articles up front, but I don't really want to maintain a complex stack either. Can this realistically be static or as close to static as possible? 
- Then there are the implications of these choices on the end user and the trade-offs we need to get to make it workable. 

One thing I'm interested in is making as much of the site/app as static as possible. When I static, I don't mean 'no server', I mean plain HTML that doesn't use javascript. We'll need some javascript but if the bulk of our content is actually plain markdown text and we only use JS for certain things like animated charts/ widgets and probably some of the general navigation/ broader UX patterns, then we don't want absolutely every piece of content to be a full component that will be server rendered + hydrated. Even with Svelte's static HTML optimisations (via `innerHTML`) and if we abuse Svelte's `{@html content}` mechanism, we could still be loading a huge amount of plain content (HTML) though javascript in certain cases which is something I'd like to avoid. It is easy to load bits of HTML from files and inject them into a page but how do we do that without losing the ability to create dynamic engaging experience when that makes the most sense? I think there is a path that can be walked here, with a few constraints and a little bit of cleverness we could automatically figure out if something is plain HTML (some Svelte basically already does) or potentially  dynamic and optimise on a page by page basis. If it were dynamic, we'd treat it as a component, if it were static we could just inject the HTML (loaded from a file)

This would require tooling support, it couldn't happen at the page level (kinda) because we'd need to write that content to HTML files and load and inject them at the right moment.

---

So yeah, I've thought about it a bunch but haven't really dug into the implementation too much. I really need to finish my current work on mdsvex which might unlock some of the capabilities I need to make this possible. But it might also be a dead end, not sure.

I'll let you know if I figure anything out though!

**Reply by ProQuestionAsker** [reply]
Ya, using Notion as a CMS isn't _perfect_, but it works pretty well. It makes it easy for me to write without worrying about formatting or figuring out how to link to multiple pages. It's all pretty straightforward. 

I've run into a bit of a problem which I'll detail first and then I'll get back to the rest of your thoughts. So, in short, I have a (very minimal) version of this working 🎉  Here's what I did: 

_(Background info: when you link to another page inline in Notion, it is currently output as a standard link)_

1. Create a remark plugin that wraps everything between two `h2` elements (including the first `h2`) in a `section` tag. Each `section` has a `data-id` attribute which is a slugified version of the header text. 
2. Create a remark plugin based on your suggested ["extract headings" plugin](https://github.com/pngwn/MDsveX/discussions/268#discussioncomment-831882) that looks specifically for a `section` with `data-id` of `summary` and injects it into the frontmatter. (Like I said, this is a minimal version, but for now it works). For now, the `summary` contents that are injected into the frontmatter are stringified using `tree_to_string = (await import('mdast-util-to-string')).toString;` **This is where the problem lies, I'm not sure how to go from MDAST to HTML for injecting into the frontmatter**
3. Create a custom `<a>` component. If the `href` was an external link, it returns a standard `<a>` link. If not, it passes its `href` and `<slot />` to another component `<Nutshell />`.
4. In the `<Nutshell />` component, I import the `.md` file based on the `href` value passed through  from`<a>`. That specific section looks something like this: 
``` js
    export let href;

    let summary;

    // if this collapsible has a link, then it should contain the summary from another page. 
    // import it
    beforeUpdate(async () => {
            if (href && !summary){
                const full = await import(`../posts/${href}/index.md`)
                const metadata = full.metadata
                summary = metadata.summary.length ? metadata.summary[0].contents : 'missing summary';
            }
    })
```
5. Then, in `<Nutshell />` I can render the `summary` content like this `<p>{summary}</p>`. 

Works like a charm! Like I alluded to earlier though, I'd rather output HTML, rather than stringified section content in Step 2 so that in Step 5, I can take advantage of that with `<p>{@html summary}</p>`. What's the best way to do that? 

Eventually I may even figure out how to expand this without putting everything in the frontmatter ;) 

---

You've clearly thought _a lot_ about this, so I'm excited to see what you come up with! I'm still new at a lot of these pieces, but happy to help out where I can if that would be helpful. 

Regarding making everything as static as possible, I think it's a great idea. Really sounds like it'll all come down to scale. How many posts are there? How many need to embedded in a single post (or can they be recursive and embed content within embedded content?) I've got a solution for my system that kinda works but I'm a little unsure if it'll scale.

**Reply by ProQuestionAsker** [reply]
Ha, silly me, `mdast-util-to-hast` and `hast-util-to-html` did what I needed 😊
